// mach/lib/headers/AcceptLanguage.js
"use strict";

var _prototypeProperties = function (child, staticProps, instanceProps) { if (staticProps) Object.defineProperties(child, staticProps); if (instanceProps) Object.defineProperties(child.prototype, instanceProps); };

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

var parseMediaValue = require("../utils/parseMediaValue");
var parseMediaValues = require("../utils/parseMediaValues");
var qualityFactorForMediaValue = require("../utils/qualityFactorForMediaValue");
var stringifyMediaValues = require("../utils/stringifyMediaValues");
var stringifyMediaValueWithoutQualityFactor = require("../utils/stringifyMediaValueWithoutQualityFactor");
var Header = require("../Header");

function byHighestPrecedence(a, b) {
  // "*" gets least precedence, all others are compared by specificity
  return a === "*" ? -1 : b === "*" ? 1 : byMostSpecific(a, b);
}

function byMostSpecific(a, b) {
  return stringifyMediaValueWithoutQualityFactor(b).length - stringifyMediaValueWithoutQualityFactor(a).length;
}

/**
 * Represents an HTTP Accept-Language header and provides several methods
 * for determining acceptable content languages.
 *
 * http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.4
 */

var AcceptLanguage = (function (Header) {
  function AcceptLanguage(value) {
    _classCallCheck(this, AcceptLanguage);

    _get(Object.getPrototypeOf(AcceptLanguage.prototype), "constructor", this).call(this, "Accept-Language", value);
  }

  _inherits(AcceptLanguage, Header);

  _prototypeProperties(AcceptLanguage, null, {
    value: {

      /**
       * Returns the value of this header as a string.
       */

      get: function () {
        return stringifyMediaValues(this._mediaValues, "-") || "";
      },
      set: function (value) {
        this._mediaValues = value ? parseMediaValues(value, "-") : [];
      },
      configurable: true
    },
    accepts: {

      /**
       * Returns true if the given language is acceptable.
       */

      value: function accepts(language) {
        return this.qualityFactorForLanguage(language) !== 0;
      },
      writable: true,
      configurable: true
    },
    qualityFactorForLanguage: {

      /**
       * Returns the quality factor for the given language.
       */

      value: function qualityFactorForLanguage(language) {
        var values = this._mediaValues;

        if (!values.length) {
          return 1;
        }var givenValue = parseMediaValue(language, "-");
        var matchingValues = values.filter(function (value) {
          if (value.type === "*") return true;

          if (value.subtype && value.subtype !== givenValue.subtype) return false;

          return value.type === givenValue.type;
        }).sort(byHighestPrecedence);

        if (!matchingValues.length) {
          return 0;
        }return qualityFactorForMediaValue(matchingValues[0]);
      },
      writable: true,
      configurable: true
    }
  });

  return AcceptLanguage;
})(Header);

module.exports = AcceptLanguage;
